### PR TITLE
Feature/chart, enhanced websocket support, dynamic generating horizontaltablist

### PR DESCRIPTION
**Pull request by gmverdon**

### CODE REVIEW

File: src/components/Chart/index.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+// Bootstrap components
+import { Card } from 'reactstrap';
+
+// libraries
+import TradingViewWidget from 'react-tradingview-widget';
+
+class Chart extends Component {
+
+  shouldComponentUpdate = (nextProps) => this.props.selectedCrypto !== nextProps.selectedCrypto
```
Comment (rafaelklaessen) [inline src/components/Chart/index.js]: Is this actually necessary? TradingViewWidget does this too.

File: src/scenes/TradeHub/index.js [hunk 2]
```diff
@@ -71,11 +114,9 @@ class TradeHub extends Component {
   }
 
   setHighestPrice(currentPrice) {
```
Comment (gmverdon) [inline src/scenes/TradeHub/index.js]: name parameter to price

File: src/components/Chart/index.js (hunk 1 above)
Comment (gmverdon) [reply src/components/Chart/index.js]: Removed

File: src/components/Chart/index.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+// Bootstrap components
+import { Card } from 'reactstrap';
+
+// libraries
+import TradingViewWidget from 'react-tradingview-widget';
+
+class Chart extends Component {
+  render() {
+    let symbol = "BINANCE:" + this.props.selectedCrypto.symbol;
```
Comment (rafaelklaessen) [inline src/components/Chart/index.js]: Use `const` instead of `let`

File: src/components/HorizontalTabList/index.js [hunk 4]
```diff
@@ -1,89 +1,32 @@
...
-  },
-  {
-    title: 'VeChain',
-  }
-];
-
 class HorizontalTabList extends Component {
+
   getListItems() {
+    let list = this.props.list;
```
Comment (rafaelklaessen) [inline src/components/HorizontalTabList/index.js]: Use `const` instead of `let`

File: src/components/HorizontalTabList/index.js [hunk 5]
```diff
@@ -1,89 +1,32 @@
...
     let listItems = [];
 
-    for (var i = 0; i < tilesData.length; i++) {
-      let newItem = this.generateListItem(tilesData[i], i)
+    for (var i = 0; i < list.length; i++) {
+      let newItem = this.generateListItem(list[i].symbol, i)
```
Comment (rafaelklaessen) [inline src/components/HorizontalTabList/index.js]: Use `const` instead of `let`

File: src/scenes/TradeHub/index.js [hunk 6]
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
...
+    this.bindSocket(newCrypto);
+  }
+
+  removeSocket(endpoint) {
+    binance.websockets.terminate(endpoint);
+  }
+
+  bindSocket = (symbol) => {
+    binance.websockets.candlesticks([symbol], "1m", (candlesticks) => {
+      let { k:ticks } = candlesticks;
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use `const` instead of `let`

File: src/scenes/TradeHub/index.js [hunk 7]
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
...
+      let { c:close } = ticks;
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use `const` instead of `let`

File: src/scenes/TradeHub/index.js [hunk 8]
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
...
       //console.log(symbol+" "+interval+" candlestick update");
       let currentPrice = parseFloat(close);
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use `const` instead of `let`

File: src/scenes/TradeHub/index.js [hunk 9]
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
...
-
-      self.setState({
+      this.setState({
         currentPrice: currentPrice
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use
```javascript
this.setState({ currentPrice });
```

File: src/scenes/TradeHub/index.js [hunk 10]
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
...
 
-  checkPrice(currentPrice) {
-    if (currentPrice > this.state.boughtPrice) {
-      if (this.isHighestPrice(currentPrice)) {
-        this.setHighestPrice(currentPrice);
+  getCryptoList() {
+    fetch("https://api.binance.com/api/v1/exchangeInfo").then((res) => {
+      return res.json();
+    }).then((data) => {
+      let cryptoList = [];
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use `const` instead of `let`

File: src/scenes/TradeHub/index.js [hunk 11]
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
...
     });
   }
 
-  checkPrice(currentPrice) {
-    if (currentPrice > this.state.boughtPrice) {
-      if (this.isHighestPrice(currentPrice)) {
-        this.setHighestPrice(currentPrice);
+  getCryptoList() {
+    fetch("https://api.binance.com/api/v1/exchangeInfo").then((res) => {
+      return res.json();
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]:
```javascript
.then(res => res.json())
```

File: src/scenes/TradeHub/index.js [hunk 12]
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
...
+    }).then((data) => {
+      let cryptoList = [];
+
+      let symbols = data.symbols;
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use `const` instead of `let`

File: src/scenes/TradeHub/index.js [hunk 13]
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
...
+      for (var i = 0; i < symbols.length; i++) {
+          cryptoList.push(symbols[i]);
+      }
+
+      if (cryptoList.length > 0) {
+        this.setState({
+          cryptoList: cryptoList,
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Remove `: cryptoList` as it's redundant

File: src/scenes/TradeHub/index.js [hunk 14]
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
...
-      if (this.isHighestPrice(currentPrice)) {
-        this.setHighestPrice(currentPrice);
+  getCryptoList() {
+    fetch("https://api.binance.com/api/v1/exchangeInfo").then((res) => {
+      return res.json();
+    }).then((data) => {
+      let cryptoList = [];
+
+      let symbols = data.symbols;
+      for (var i = 0; i < symbols.length; i++) {
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use `let` instead of `var`

File: src/scenes/TradeHub/index.js [hunk 15]
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
...
+        this.sell(price);
         return;
       }
     }
   }
 
-  isHighestPrice(currentPrice) {
-    let isHighestPrice = currentPrice > this.state.highestPrice;
+  isHighestPrice(price) {
+    let isHighestPrice = price > this.state.highestPrice;
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use `const` instead of `let`

File: src/scenes/TradeHub/index.js [hunk 16]
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
...
+      highestPrice: price
     })
-    console.log("New highestPrice: " + this.state.highestPrice)
   }
 
-  shouldSell(currentPrice) {
+  shouldSell(price) {
     if (this.state.sellEnabled && !this.state.hasSold) {
-      let shouldSell = (currentPrice <= (this.state.highestPrice - this.state.highestPrice * this.state.diffPercentage));
+      let shouldSell = (price <= (this.state.highestPrice - this.state.highestPrice * this.state.diffPercentage));
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use `const` instead of `let`

File: src/scenes/TradeHub/index.js [hunk 17]
```diff
@@ -112,22 +152,44 @@ class TradeHub extends Component {
     })
   }
 
+  changeSelectedCrypto(symbol) {
+    let crypto = this.state.cryptoList.find((obj) => { return obj.symbol === symbol; });
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use `const` instead of `let`

File: src/scenes/TradeHub/index.js [hunk 18]
```diff
@@ -112,22 +152,44 @@ class TradeHub extends Component {
...
+    if (crypto == null) { return; }
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use
```javascript
if (crypto === null) return;
```

File: src/scenes/TradeHub/index.js [hunk 19]
```diff
@@ -112,22 +152,44 @@ class TradeHub extends Component {
...
+
+    binance.prices((ticker) => {
+      this.setState({
+        currentPrice: parseFloat(ticker[crypto.symbol])
+      })
+    });
+
+    this.setState({
+      selectedCrypto: crypto
+    },() => { this.rebindSocket(); });
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: () => this.rebindSocket(); is enough

File: src/scenes/TradeHub/index.js [hunk 20]
```diff
@@ -112,22 +152,44 @@ class TradeHub extends Component {
...
+
   render() {
     let selectedCrypto = this.state.selectedCrypto;
-    let currentPrice = this.state.currentPrice.toFixed(2);
-    let highestPrice = this.state.highestPrice.toFixed(2);;
-    let sellPrice = (highestPrice - highestPrice * this.state.diffPercentage).toFixed(2);
+    let currentPrice = this.state.currentPrice.toFixed(6);
+    let highestPrice = this.state.highestPrice.toFixed(6);;
+    let sellPrice = (highestPrice - highestPrice * this.state.diffPercentage).toFixed(6);
+    let cryptoList = this.state.cryptoList;
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use `const` instead of `let`

File: src/scenes/TradeHub/index.js [hunk 21]
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
...
       }
     }
   }
 
-  isHighestPrice(currentPrice) {
-    let isHighestPrice = currentPrice > this.state.highestPrice;
+  isHighestPrice(price) {
+    const isHighestPrice = price > this.state.highestPrice;
     return isHighestPrice;
   }
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Also possible:
```javascript
isHighestPrice = price => price > this.state.highestPrice;
```

File: src/scenes/TradeHub/index.js [hunk 22]
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
...
 
-  shouldSell(currentPrice) {
+  shouldSell(price) {
     if (this.state.sellEnabled && !this.state.hasSold) {
-      let shouldSell = (currentPrice <= (this.state.highestPrice - this.state.highestPrice * this.state.diffPercentage));
+      const shouldSell = (price <= (this.state.highestPrice - this.state.highestPrice * this.state.diffPercentage));
       return shouldSell;
     }
   }
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: How about using object extraction here?
```javascript
const { sellEnabled, hasSold, highestPrice, diffPercentage } = this.state;
if (sellEnabled && !hasSold) {
  return price <= highestPrice - highestPrice * diffPercentage;
}
```

File: src/scenes/TradeHub/index.js [hunk 23]
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
...
 
-  sell(currentPrice) {
-    alert("SOLD at " + currentPrice);
-    console.log("SOLD at " + currentPrice);
+  sell(price) {
+    alert("SOLD at " + price);
     this.setState({
       hasSold: true
     })
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Add ;

File: src/scenes/TradeHub/index.js [hunk 24]
```diff
@@ -100,7 +134,7 @@ class TradeHub extends Component {
   }
 
   setDiffPercentage(e) {
-    let diffPercentage = (e.target.value / 100);
+    const diffPercentage = (e.target.value / 100);
     this.setState({
       diffPercentage: diffPercentage
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Can be written as
```javascript
this.setState({ diffPercentage });
```

File: src/scenes/TradeHub/index.js [hunk 25]
```diff
@@ -112,22 +146,46 @@ class TradeHub extends Component {
...
   }
 
+  changeSelectedCrypto(symbol) {
+    const crypto = this.state.cryptoList.find((obj) => { return obj.symbol === symbol; });
+    if (crypto === null) return;
+
+    binance.prices((ticker) => {
+      const currentPrice = parseFloat(ticker[crypto.symbol]);
+      this.setState({
+        currentPrice: currentPrice,
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: `currentPrice` twice is redundant

File: src/scenes/TradeHub/index.js [hunk 26]
```diff
@@ -112,22 +146,46 @@ class TradeHub extends Component {
     })
   }
 
+  changeSelectedCrypto(symbol) {
+    const crypto = this.state.cryptoList.find((obj) => { return obj.symbol === symbol; });
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use
```javascript
this.state.cryptoList.find(obj => obj.symbol === symbol);
```

File: src/scenes/TradeHub/index.js [hunk 27]
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
...
-  checkPrice(currentPrice) {
-    if (currentPrice > this.state.boughtPrice) {
-      if (this.isHighestPrice(currentPrice)) {
-        this.setHighestPrice(currentPrice);
+  getCryptoList() {
+    fetch("https://api.binance.com/api/v1/exchangeInfo").then(res => res.json()).then((data) => {
+      const cryptoList = [];
+      for (let i = 0; i < data.symbols.length; i++) {
+          cryptoList.push(data.symbols[i]);
+      }
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use
```javascript
const cryptoList = data.symbols;
```

File: src/scenes/TradeHub/index.js [hunk 28]
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
...
+      this.checkPrice(currentPrice);
     });
   }
 
-  checkPrice(currentPrice) {
-    if (currentPrice > this.state.boughtPrice) {
-      if (this.isHighestPrice(currentPrice)) {
-        this.setHighestPrice(currentPrice);
+  getCryptoList() {
+    fetch("https://api.binance.com/api/v1/exchangeInfo").then(res => res.json()).then((data) => {
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Use single quotes

File: src/scenes/TradeHub/index.js [hunk 29]
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
...
+      if (this.isHighestPrice(price)) {
+        this.setHighestPrice(price);
         return;
       }
 
-      if (this.shouldSell(currentPrice)) {
-        this.sell(currentPrice);
+      if (this.shouldSell(price)) {
+        this.sell(price);
         return;
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js]: Unnecessary return statement

File: src/scenes/TradeHub/index.js (hunk 27 above)
Comment (gmverdon) [reply src/scenes/TradeHub/index.js]: Nice one!

File: src/components/Chart/index.js [hunk 30]
```diff
@@ -0,0 +1,24 @@
+import React, { Component } from 'react';
+import './styles.css';
+
+// Bootstrap components
+import { Card } from 'reactstrap';
+
+// libraries
+import TradingViewWidget from 'react-tradingview-widget';
+
+class Chart extends Component {
```
Comment (rafaelklaessen) [inline src/components/Chart/index.js]: Use a functional component?

File: src/components/Chart/index.js [hunk 31]
```diff
@@ -0,0 +1,24 @@
+import React, { Component } from 'react';
+import './styles.css';
+
+// Bootstrap components
```
Comment (rafaelklaessen) [inline src/components/Chart/index.js]: That's obvious

File: src/scenes/TradeHub/index.js [hunk 32]
```diff
@@ -8,126 +8,178 @@ import binance from 'binance-api';
...
-        this.sell(currentPrice);
+  checkPrice(price) {
+    if (price > this.state.boughtPrice) {
+      if (this.isHighestPrice(price)) {
+        this.setHighestPrice(price);
         return;
       }
+
+      if (this.shouldSell(price)) {
+        this.sell(price);
```
Comment (rafaelklaessen) [inline src/scenes/TradeHub/index.js:95]: You *could* do
```javascript
if (this.shouldSell(price)) this.sell(price):
```
Without the curly braces

**Pull request merged**